networkit/cpp/centrality/GroupHarmonicCloseness.cpp

```diff
@@ -0,0 +1,695 @@
+                if (candidateNodesPQ.empty()) {
+                    stop.store(true, std::memory_order_relaxed);
+                } else {
+                    u = candidateNodesPQ.extract_top();
+                    if (margGain[u] <= bestScore) {
+                        stop.store(true, std::memory_order_relaxed);
+                        u = none;
+                    }
+                }
+            }
+
+            if (u == none)
+                break;
+
+            const auto ssspResult = prunedSSSP(u, bestScore);
+            margGain[u] = ssspResult.score;
+
+#pragma omp critical
+            {
+                if (ssspResult.complete && ssspResult.score > bestScore) {
+                    bestNode = u;
+                    bestScore = ssspResult.score;
+                    std::swap(curDistGlobal[omp_get_thread_num()], curBestDist);
+                    std::swap(curNearerNodesGlobal[omp_get_thread_num()], nearerNodes);
+                }
+            }
+        }
+    }
+
+    assert(bestNode != none);
+    for (const node u : nearerNodes) {
+        assert(distFromGroup[u] > curBestDist[u]);
+        distFromGroup[u] = curBestDist[u];
+        if (!G->isWeighted() && !G->isDirected() && distFromGroup[u] > 1 && curBestDist[u] <= 1) {
+            assert(reachableNodesInComponent[component->subsetOf(u)]);
+            --reachableNodesInComponent[component->subsetOf(u)];
+        }
+    }
+
+    return bestNode;
+}
+
+template <>
+double GroupHarmonicClosenessImpl<count>::scoreOfGroup(const Graph &graph,
+                                                       const std::vector<node> &inputGroup) {
+    double score = 0.;
+    Traversal::BFSfrom(graph, inputGroup.begin(), inputGroup.end(),
+                       [&score](node, edgeweight dist) {
+                           if (dist > 0.)
+                               score += 1. / dist;
+                       });
+    return score;
+}
+
+template <>
+double GroupHarmonicClosenessImpl<edgeweight>::scoreOfGroup(const Graph &graph,
+                                                            const std::vector<node> &inputGroup) {
+    double score = 0.;
+    Traversal::DijkstraFrom(graph, inputGroup.begin(), inputGroup.end(),
+                            [&score](node, edgeweight dist) {
+                                if (dist > 0.)
+                                    score += 1. / dist;
+                            });
+    return score;
+}
+
+#ifdef NETWORKIT_SANITY_CHECKS
+template <>
+void GroupHarmonicClosenessImpl<count>::checkDistFromGroup() const {
+    Traversal::BFSfrom(
+        *G, group.begin(), group.end(),
+        [&distFromGroup = distFromGroup](node x, count d) { assert(distFromGroup[x] == d); });
+}
+
+template <>
+void GroupHarmonicClosenessImpl<edgeweight>::checkDistFromGroup() const {
+    Traversal::DijkstraFrom(
+        *G, group.begin(), group.end(),
+        [&distFromGroup = distFromGroup](node x, edgeweight d) { assert(distFromGroup[x] == d); });
+}
```

Comment:
I wonder how much of these specializations could be avoided by specializing a `traversalFromGroup` function instead and then using that function in a non-specialized `checkDistFromGroup` (and other functions?).